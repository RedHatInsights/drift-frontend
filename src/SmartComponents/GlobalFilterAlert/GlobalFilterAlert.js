import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { Alert, AlertActionCloseButton } from '@patternfly/react-core';

export class GlobalFilterAlert extends Component {
    constructor(props) {
        super(props);

        this.state = {
            isOpen: true
        };

        this.toggleIsOpen = () => {
            const { isOpen } = this.state;

            this.setState({
                isOpen: !isOpen
            });
        };
    }

    buildBody = () => {
        const { sidsFilter, tagsFilter, workloadsFilter } = this.props.globalFilterState;
        let filters = 'Workloads: ' + Object.keys(workloadsFilter)[0] + '. ';

        if (sidsFilter.length) {
            filters += 'SAP ID (SID): ';
            for (let i = 0; i < sidsFilter.length; i++) {
                filters += sidsFilter[i];
                if (i + 1 === sidsFilter.length) {
                    filters += '. ';
                } else {
                    filters += ', ';
                }
            }
        }

        if (tagsFilter.length) {
            let tags = [];
            let tagsList = {};
            filters += 'Tags: ';

            tagsFilter.forEach(function(tag) {
                tags.push(tag.split('/'));
            });

            tags.forEach(function(tag) {
                if (!(tag[0] in tagsList)) {
                    tagsList[tag[0]] = [ tag[1] ];
                } else {
                    tagsList[tag[0]].push(tag[1]);
                }
            });

            for (const [ key, value ] of Object.entries(tagsList)) {
                filters += key + ': ';
                for (let i = 0; i < value.length; i++) {
                    filters += value[i];
                    if (i + 1 === value.length) {
                        filters += '. ';
                    } else {
                        filters += ', ';
                    }
                }
            }
        }

        return filters;
    }

    render() {
        const { sidsFilter, tagsFilter, workloadsFilter } = this.props.globalFilterState;
        const { isOpen } = this.state;

        return (
            <React.Fragment>
                { isOpen && (workloadsFilter.SAP?.isSelected || sidsFilter.length > 0 || tagsFilter.length > 0)
                    ? <Alert
                        variant='info'
                        title='Your systems are pre-filtered by the global context selector.'
                        isInline
                        actionClose={ <AlertActionCloseButton onClose={ () => this.toggleIsOpen() } /> }
                    >
                        <p>
                            { this.buildBody() }
                        </p>
                    </Alert>
                    : null
                }
            </React.Fragment>
        );
    }
}

GlobalFilterAlert.propTypes = {
    globalFilterState: PropTypes.object
};

export default GlobalFilterAlert;
