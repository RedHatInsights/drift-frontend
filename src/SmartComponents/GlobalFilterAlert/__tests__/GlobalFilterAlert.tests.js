import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';

import { GlobalFilterAlert } from '../GlobalFilterAlert';

describe('GlobalFilterAlert', () => {
    let props;

    beforeEach(() => {
        props = {
            globalFilterState: {
                sidsFilter: [],
                tagsFilter: [],
                workloadsFilter: {}
            }
        };
    });

    it('should not render when empty', () => {
        const wrapper = shallow(
            <GlobalFilterAlert
                { ...props }
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should not render with all workloads', () => {
        props.globalFilterState.workloadsFilter = {
            'All Workloads': {
                group: {
                    name: 'Workloads'
                }
            }
        };

        const wrapper = shallow(
            <GlobalFilterAlert
                { ...props }
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with SAP', () => {
        props.globalFilterState.workloadsFilter = {
            SAP: {
                group: {
                    name: 'Workloads'
                },
                item: {
                    value: 'SAP'
                },
                isSelected: true
            }
        };

        const wrapper = shallow(
            <GlobalFilterAlert
                { ...props }
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should NOT render with not selected SAP', () => {
        props.globalFilterState.workloadsFilter = {
            SAP: {
                group: {
                    name: 'Workloads'
                },
                item: {
                    value: 'SAP'
                },
                isSelected: false
            }
        };

        const wrapper = shallow(
            <GlobalFilterAlert
                { ...props }
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });

    it('should render with SAP and filters', () => {
        props.globalFilterState.workloadsFilter = {
            SAP: {
                group: {
                    name: 'Workloads'
                },
                item: {
                    value: 'SAP'
                },
                isSelected: true
            }
        };

        props.globalFilterState.sidsFilter = [ 'AB1', 'XY1' ];
        props.globalFilterState.tagsFilter = [ 'patch/rest=patchman-engine', 'patch/dev=patchman-engine', 'insights-client/group=XmygroupX' ];

        const wrapper = shallow(
            <GlobalFilterAlert
                { ...props }
            />
        );

        expect(toJson(wrapper)).toMatchSnapshot();
    });
});
