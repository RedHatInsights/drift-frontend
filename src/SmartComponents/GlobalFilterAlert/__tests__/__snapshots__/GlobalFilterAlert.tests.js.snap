// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GlobalFilterAlert should not render when empty 1`] = `<Fragment />`;

exports[`GlobalFilterAlert should not render with all workloads 1`] = `<Fragment />`;

exports[`GlobalFilterAlert should render with SAP 1`] = `
<Fragment>
  <Alert
    actionClose={
      <AlertActionCloseButton
        onClose={[Function]}
      />
    }
    isInline={true}
    title="Your systems are pre-filtered by the global context selector."
    variant="info"
  >
    <p>
      Workloads: SAP. 
    </p>
  </Alert>
</Fragment>
`;

exports[`GlobalFilterAlert should render with SAP and filters 1`] = `
<Fragment>
  <Alert
    actionClose={
      <AlertActionCloseButton
        onClose={[Function]}
      />
    }
    isInline={true}
    title="Your systems are pre-filtered by the global context selector."
    variant="info"
  >
    <p>
      Workloads: SAP. SAP ID (SID): AB1, XY1. Tags: patch: rest=patchman-engine, dev=patchman-engine. insights-client: group=XmygroupX. 
    </p>
  </Alert>
</Fragment>
`;
